python/lsst/obs/decam/decamCpIsr.py

```diff
@@ -0,0 +1,145 @@
+#
+# LSST Data Management System
+# Copyright 2016 LSST Corporation.
+#
+# This product includes software developed by the
+# LSST Project (http://www.lsst.org/).
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the LSST License Statement and
+# the GNU General Public License along with this program.  If not,
+# see <http://www.lsstcorp.org/LegalNotices/>.
+#
+
+from lsst.ip.isr import biasCorrection, flatCorrection
+from lsst.meas.algorithms.detection import SourceDetectionTask
+import lsst.pipe.base as pipeBase
+from .isr import DecamIsrTask
+
+
+def _computeEdgeSize(rawExposure, calibExposure):
+    """Compute the number of edge trim pixels of the calibration product.
+    Some Community Pipeline Calibration products are trimmed on their edges
+    and are smaller than the raw data. Use the dimension difference between
+    raw exposure and the calibration product to compute the edge trim pixels.
+    @param[in] rawExposure: the data section of a raw exposure
+    @param[in] calibExposure: calibration bias or flat exposure,
+                              known to be smaller than raw data
+    @return an integer as the number of trimmed pixels on each edge
+    """
+    nx, ny = rawExposure.getBBox().getDimensions() - calibExposure.getBBox().getDimensions()
+    assert nx == ny, "Exposure is trimmed differently in X and Y"
+    assert nx%2 == 0, "Exposure is trimmed unevenly in X"
+    assert nx >= 0, "Calibration image is larger than raw data"
+    return nx/2
+
+
+class DecamCpIsrTask(DecamIsrTask):
+    """Perform ISR task using Community Pipeline Calibration Products MasterCal
+
+    The CP MasterCal products have butler dataset types cpBias and cpFlat,
+    different from the LSST-generated calibration products (bias/flat).
+    """
+    def readIsrData(self, dataRef, rawExposure):
```

Comment:
I've not considered it very hard. My first thought was that most of the duplication could be avoided by using attributes to define the dataset types rather than hard-coding them in the method. Thus something like:

```python
class IsrTask(...)
    biasDatasetType = "bias"
    [...]
    def readIsrData(...):
        biasExposure = self.getIsrExposure(dataRef, self.biasDatasetType)
        [...]

class DecamCpIsrTask(IsrTask):
    biasDatasetType = "cpBias"
```

(Needs a bit more imagination to deal with fringes.)

Whether that would be any better (or worse) than changing the mapper I'm not sure.